### ISSUE TITLE
Recursive includes that are added may not be found for open files (until the IntelliSense process gets restarted)

### ISSUE THREAD

**Issue by taronlee** [microsoft/vscode-cpptools#7783]
Bug type: Language Service

**Describe the bug**
- OS and Version: windows 10
- VS Code Version: 1.57.1
- C/C++ Extension Version: 1.4.1
- Other extensions you installed (and if the issue persists after disabling them): remote-ssh
- If using SSH remote, specify OS of remote machine: linux
- A clear and concise description of what the bug is, including information about the workspace (i.e. is the workspace a single project or multiple projects, size of the project, etc).

**we want to provide the IntelliSense automatically.**
our project is that when user open the project with VS Code, collecting all of the header files and providing IntelliSense automatically. (processing in the background)

so we create the c_cpp_properties.json using the `fs-extra` module.
=> not using the `c/c++: Edit Configuration (JSON)`

but if one of the paths in includePath is not in `vscode.workspace.workspaceFolder` path,
IntelliSense doesn't work automatically.

1. **not working**
/home3/the.lee/
-----------------**/.test/headers**
-----------------/project/worksapce
-------------------------/.vscode/c_cpp_ properties.json
```javascript
"includePath": [
                "/home3/the.lee/.test/headers/**",
                "${workspaceFolder}/**"
],
```
=> intelliSense doesn't work **until relaunch**

2. **working**
/home3/the.lee/
-----------------/project/worksapce
-------------------------/.vscode/c_cpp_ properties.json
-------------------------**/.test/headers**
```javascript
"includePath": [
                "${workspaceFolder}/**"
],
```
=> it works automatically.

`and is there a way to activate 'searching include path...' behavior whenever I want?`

**Steps to reproduce**

1. create 'c_cpp_properties.json' using `fs` module (not c/c++: Edit Configuration)
```javascript
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "/home3/the.lee/.test/headers/**",
                "${workspaceFolder}/**"
            ],
            "defines": [],
            "compilerPath": "/usr/bin/clang-5.0",
            "cStandard": "c11",
            "cppStandard": "c++98",
            "intelliSenseMode": "linux-clang-x64"
        }
    ],
    "version": 4
}

```

**Expected behavior**

IntelliSense should work automatically even without relaunch.

**Code sample and logs**

- Code sample

- Configurations in `c_cpp_properties.json`

- Logs from running `C/C++: Log Diagnostics` from the VS Code command palette

- Logs from [the language server logging](https://code.visualstudio.com/docs/cpp/enable-logging-cpp#_enable-logging-for-the-language-server)

**Screenshots**

**Additional context**

**Comment by sean-mcmanus** [comment]
The problem is that the IntelliSense updating is being done too soon before the recursive includes are processed. One workaround is to close all files before generating the c_cpp_properties.json. Another workaround is to run C/C++: Rescan Workspace from the command palette.

**Comment by sean-mcmanus** [comment]
Using setting
```
    "C_Cpp.files.exclude": {
        "**/.vscode": false
    }
```
is another workaround.

**Comment by taronlee** [comment]
Thank you for your answer.
```
Another workaround is to run C/C++: Rescan Workspace from the command palette.
Another workaround is to switch configurations (and adding a duplicate to switch to if a 2nd configuration doesn't exist).
```
I will test those workarounds and tell you about the result.

**Comment by sean-mcmanus** [comment]
Also setting C_Cpp.intelliSenseUpdateDelay to the default of 2000 (the default) or 3000 may also cause the issue to not repro (if you changed it to 500)....depending on your workspace size.

**Comment by sean-mcmanus** [comment]
We have a fix, but it probably won't be available till 1.6.0-insiders, since it's a little too risky to add to 1.5.0.

**Comment by taronlee** [comment]
Got it.
And also when I excuted `C/C++: Rescan Workspace`, it's not working.

**Comment by sean-mcmanus** [comment]
Hmm, C/C++: Rescan Workspace works for me. Your logging should show something like
```
cpptools/rescanFolder
Shutting down IntelliSense server: <path>/TEST.CPP
Populate include completion cache.
Discovering files...
...
Checking for syntax errors: file:<path>/test.cpp
Done parsing open files.
sending compilation args for <path>\TEST.CPP
  include: <recursive include directory>
```

The fix should be available with 1.6.0-insiders (and later 1.6.0), but it's possibly you may be hitting some additional issue.

**Comment by taronlee** [comment]
> Hmm, C/C++: Rescan Workspace works for me. Your logging should show something like
> 
> ```
> cpptools/rescanFolder
> Shutting down IntelliSense server: <path>/TEST.CPP
> Populate include completion cache.
> Discovering files...
> ...
> Checking for syntax errors: file:<path>/test.cpp
> Done parsing open files.
> sending compilation args for <path>\TEST.CPP
>   include: <recursive include directory>
> ```
> 
> The fix should be available with 1.6.0-insiders (and later 1.6.0), but it's possibly you may be hitting some additional issue.

in case of <path>/TEST.cpp,
is the `<path>` external path? isn't it? (not in workspace path in the project)

**Comment by sean-mcmanus** [comment]
I tried it with `<path>/test.cpp` in the workspace folder and the header (`<recursive include directory>`) being external. I'll try both being external tomorrow...

**Comment by taronlee** [comment]
a

> I tried it with `<path>/test.cpp` in the workspace folder and the header (`<recursive include directory>`) being external. I'll try both being external tomorrow...

oh.. sorry,
I meant the environment you said..
I will try again.

**Comment by taronlee** [comment]
I always appreciate your taking the time to do this.

`C/C++: Rescan Workspace`
As a result, It works.

But it should be called more than 2 times.
I've tried it many times, but it won't work if it's only called once..

Are there any rules? Need to call after several mins?

**Comment by sean-mcmanus** [comment]
If the 1st Rescan doesn't work, it's probably because it's been done too soon and the recursive include directory hasn't been enumerated yet. You can verify this by setting C_Cpp.loggingLevel to "Debug" and looking for logging "Done discovering files."...you'll want to wait till after that happens. Doing the Rescan should cause the IntelliSense process with the missing include path to shutdown and restart and find the recursive include the 2nd time. Our pending fix should wait for the discovering files to finish before it tries to search the database for recursive includes resolution.

**Comment by Colengms** [comment]
This should be addressed in 1.6.0-insiders.
